packages/firebase_messaging/ios/Classes/FLTFirebaseMessagingPlugin.m

```diff
@@ -278,4 +342,81 @@ - (void)messaging:(FIRMessaging *)messaging
   [_channel invokeMethod:@"onMessage" arguments:remoteMessage.appData];
 }
 
+- (void)setupBackgroundHandling:(int64_t)handle {
+  NSLog(@"Setting up Firebase background handling");
+
+  [self saveCallbackHandle:backgroundSetupCallback handle:handle];
+
+  NSLog(@"Finished background setup");
+}
+
+- (void)startBackgroundRunner {
+  NSLog(@"Starting background runner");
+
+  int64_t handle = [self getCallbackHandle:backgroundSetupCallback];
+
+  FlutterCallbackInformation *info = [FlutterCallbackCache lookupCallbackInformation:handle];
+  NSAssert(info != nil, @"failed to find callback");
+  NSString *entrypoint = info.callbackName;
+  NSString *uri = info.callbackLibraryPath;
+
+  [_headlessRunner runWithEntrypoint:entrypoint libraryURI:uri];
+  [_registrar addMethodCallDelegate:self channel:_backgroundChannel];
+
+  // Once our headless runner has been started, we need to register the application's plugins
+  // with the runner in order for them to work on the background isolate. `registerPlugins` is
+  // a callback set from AppDelegate.m in the main application. This callback should register
+  // all relevant plugins (excluding those which require UI).
+
+  NSAssert(registerPlugins != nil, @"failed to set registerPlugins");
+  registerPlugins(_headlessRunner);
+}
+
+- (int64_t)getCallbackHandle:(NSString *) key {
+  NSLog(@"Getting callback handle for key %@", key);
+  id handle = [_userDefaults objectForKey:key];
+  if (handle == nil) {
+      return 0;
+  }
+  return [handle longLongValue];
+}
+
+- (void) saveCallbackHandle:(NSString *)key handle:(int64_t)handle {
+  NSLog(@"Saving callback handle for key %@", key);
+
+  [_userDefaults setObject:[NSNumber numberWithLongLong:handle] forKey:key];
+}
+
+- (void) queueMethodCall:(NSString *) method callbackName:(NSString*)callback arguments:(NSDictionary*)arguments {
+  NSLog(@"Queuing method call: %@", method);
+  int64_t handle = [self getCallbackHandle:callback];
+
+  @synchronized(self) {
+    if (initialized) {
+      [self invokeMethod:method callbackHandle:handle arguments:arguments];
+    } else {
+      NSArray *call = @[method, @(handle), arguments];
+      [_eventQueue addObject:call];
+    }
+  }
+}
+
+- (void) invokeMethod:(NSString *) method callbackHandle:(long)handle arguments:(NSDictionary*)arguments {
+  NSLog(@"Invoking method: %@", method);
+  //NSArray* args = @[@(handle), arguments];
+  
+  NSDictionary *aaa = @{
```

Comment:
Well, I don't understand obj-c or this code well enough to comment anything good. But if we're looking for a PR to the main repo, this variable should be renamed, the comment on 406 deleted, and all the NSLog removed.
Otherwise, 5/5 for our purpose